src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24075,6 +24075,151 @@ bool Init()
+                );
+        }
+
+        [Fact]
+        public void MemberNotNullWhenTrue_EnforcedInMethodBody_NonConstant_ButAnalyzable()
+        {
+            var c = CreateNullableCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1;
+    public string field3;
+
+    C()
+    {
+        if (!Init()) throw null!;
+    }
+
+    [MemberNotNullWhen(true, nameof(field1), nameof(field3))]
+    bool Init()
+    {
+        bool b = true;
+        if (b)
+        {
+            return field1 == null; // 1
+        }
+        if (b)
+        {
+            return field1 != null;
+        }
+        if (b)
+        {
+            return Init();
+        }
+        return !Init(); // 2, 3
+    }
+
+    bool M([MaybeNullWhen(true)]out string s) => throw null!;
+}
+", MemberNotNullWhenAttributeDefinition, MaybeNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (19,13): warning CS8775: Member 'field1' must have a non-null value when exiting with 'true'.
+                //             return field1 == null; // 1
+                Diagnostic(ErrorCode.WRN_MemberNotNullWhen, "return field1 == null;").WithArguments("field1", "true").WithLocation(19, 13),
+                // (29,9): warning CS8775: Member 'field1' must have a non-null value when exiting with 'true'.
+                //         return !Init(); // 2, 3
+                Diagnostic(ErrorCode.WRN_MemberNotNullWhen, "return !Init();").WithArguments("field1", "true").WithLocation(29, 9),
+                // (29,9): warning CS8775: Member 'field3' must have a non-null value when exiting with 'true'.
+                //         return !Init(); // 2, 3
+                Diagnostic(ErrorCode.WRN_MemberNotNullWhen, "return !Init();").WithArguments("field3", "true").WithLocation(29, 9)
+                );
+        }
+
+        [Fact]
+        public void MemberNotNullWhenFalse_EnforcedInMethodBody_NonConstant_ButAnalyzable()
+        {
+            var c = CreateNullableCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1;
+    public string field3;
+
+    C()
+    {
+        if (Init()) throw null!;
+    }
+
+    [MemberNotNullWhen(false, nameof(field1), nameof(field3))]
+    bool Init()
+    {
+        bool b = true;
+        if (b)
+        {
+            return field1 == null;
+        }
+        if (b)
+        {
+            return field1 != null; // 1
+        }
+        if (b)
+        {
+            return Init();
+        }
+        return !Init(); // 2, 3
+    }
+
+    bool M([MaybeNullWhen(true)]out string s) => throw null!;
+}
+", MemberNotNullWhenAttributeDefinition, MaybeNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (23,13): warning CS8775: Member 'field1' must have a non-null value when exiting with 'true'.
```

Comment:
Looks like this diagnostic message is wrong, maybe due to passing the wrong `sense` value to `enforceMemberNotNullWhenIfAffected`?

Reply:
Indeed, thanks